### ISSUE TITLE
Accessibility Audit

### ISSUE THREAD

**Issue by noopkat** [noopkat/electric-io#1]
1. Explore focus management
2. ✅ Keyboard access for drag and drop needs to be implemented
3. Create html landmarks for page
4. ✅ Run Axe Core on the front-end
5. fix any issues

**Comment by escherina** [comment]
Although the issue mentions `axe-core`, I've used Microsoft's [Accessibility Insights](https://chrome.google.com/webstore/detail/accessibility-insights-fo/pbjjkligggfmakdaogkfomddhfmpjeni/) extension since it's built on top of the `axe-core` engine, and I think it's one of the best accessibility tools around right now.

I ran the extension's FastPass mode which includes automated checks, and a tab stops test (to see where the focus goes on the page).

## Automated checks (showing only failures)

**[color-contrast](https://dequeuniversity.com/rules/axe/3.2/color-contrast)**: Elements must have sufficient color contrast (5) [WCAG 1.4.3](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html)
Failed elements:
* `<input type="submit" value="save" class="action-button">`
* `<input type="submit" value="create" class="action-button">`
* `<button class="action-button"> MxChip </button>`
* `<button class="action-button"> Click Me </button>`
* `<p class="status success">done!</p>`

**[html-has-lang](https://dequeuniversity.com/rules/axe/3.2/html-has-lang)**: `<html>` element must have a `lang` attribute (1) [WCAG 3.1.1](https://www.w3.org/WAI/WCAG21/Understanding/language-of-page.html)
Failed elements:
* `<html>`

**[image-alt](https://dequeuniversity.com/rules/axe/3.2/image-alt)**: Images must have alternate text (2) [WCAG 1.1.1](https://www.w3.org/WAI/WCAG21/Understanding/non-text-content.html)
Failed elements:
* `<img src="https://upload.wikimedia.org/wikipedia/commons/3/3a/Cat03.jpg" width="100%" blocksize="250,200" messages="">`
* `<img src="https://media.giphy.com/media/1wXeLxuTVBZe0Ht7Zu/giphy.gif" width="100%" blocksize="250,200" messages="">`

## Tab stops
Diagram:
![image](https://user-images.githubusercontent.com/14368631/57583263-03a97400-74cf-11e9-95a1-31d897ffa301.png)

For the screenshot I stopped tabbing after two cards so the image is readable, but in testing I tabbed all the way through.

While testing the tab order I found:
* we don't get stuck or trapped when tabbing
* once we start tabbing through the cards, the focus order goes all over the page because the cards can be rearranged to your liking. I don't see this as a problem given this is a drag-and-drop dashboard, but I did find that the focus styles were hard to spot.

## Action points

Here's what I think we should implement as a result of the audit:
* Improve colour contrast for the `.action-button` class
* Improve colour contrast for the `.success` class
* Improve visibility of focus styles
* Set a default language of `en` on `<html>`
  (I considered if this should be `en`, since users can set card titles to whatever they like, but we do have hard-coded English text in many places on the page e.g. 'Settings', 'edit', 'New card' etc)
* Add an `alt` field to the `Sticker` card type.
  Although users of the dashboard are likely to upload their own images, I do see a case where a user with a screen reader could be viewing a dashboard created by someone else. It also would provide a fallback in case the image didn't load.

Additional things to implement as per the original issue text:
* Keyboard access for drag and drop
* Create html landmarks

And in addition, I think we could follow up with screen reader testing, especially focusing on:
* How are live data updates conveyed to the user? (Would it be possible to announce updates e.g. with `aria-live` when a particular card has focus?)
* Are errors (e.g. errors in device methods) announced to the user correctly?

**Comment by escherina** [comment]
Forgot to say, I would be happy to follow up and add these fixes! 😊

**Comment by kleinfreund** [comment]
Some of the mentioned issues are rather low-hanging fruites. We could create separate issues and mark them as “good first issue” and point to these when people ask about how they could go about starting to contribute to open source. Maybe some people on stream today are interested, too.

**Comment by noopkat** [comment]
@escherina this is incredible work, and as usual this issue was written up in an exemplary manner ✨ 🌈 

I love @kleinfreund 's idea of breaking some of these up. The language attribute on the html tag is an especially good example of this.

This way, we can also take some burden off @escherina as they generously offered to do a bunch of these and they wouldn't feel obligated to complete all of them in order to close this issue.

Thank you so much again.

![bitmoji](https://render.bitstrips.com/v2/cpanel/4714cd12-a2df-401e-9448-a95fdf46790b-8016a622-e09b-4505-8856-7726a074b1f9-v1.png?transparent=1&palette=1&width=246)

**Comment by noopkat** [comment]
One 🆒  idea that just came from a11y camp seattle today was to do with the tab ordering:

We could render each card in the dom order tied to the position of the card as saved in the dashboard configuration. So left and top cards would be higher in the dom order than a card in the bottom right corner of the dashboard.

**Comment by noopkat** [comment]
FYI: fixed success text colour contrast via 46fa296cbb4607a77c00c8a43b85a40079b3f459

**Comment by kleinfreund** [comment]
What exactly are we talking about when talking about HTML landmarks? I don’t know if we use any elements in a way that require additional landmark roles. One thing we could do is make the dashboard div the `main` element as it contains the main content: the dashboard.

**Comment by mtheoryx** [comment]
I suspected we may approach a certain point in this issue where we implement 80% of the list, and then start wondering about the validity of some of the other items. It feels like we may be at that point regarding landmarks.

While it was an amazing aggregation of a lot of a11y improvements, generously put together in a nice list with plenty of context, I would be open to moving the landmark topic to it's own issue for separate evaluation and effort.

Thoughts?

**Comment by noopkat** [comment]
fixed the last of the fast pass failures with 688ee5b2c96c1ed6f4853cc0b092667c8facbd46

🙇‍♀

**Comment by noopkat** [comment]
Agree with everyone about closing this issue in favour of opening new ones for each remaining a11y enhancement. Closing this now and pouring one out ❤️
